File: ido_sdk_flutter/ido_sdk/protocol_sdk/protocol_channel/android/src/main/kotlin/com/idosmart/model/IDOEmotionHealthReminderModel.kt
//
//  IDOEmotionHealthReminderModel.kt
//  protocol_channel
//
//  Created by hc on 2025/12/04.
//

package com.idosmart.model

import com.google.gson.*
import com.google.gson.annotations.SerializedName
import java.io.Serializable
import java.lang.reflect.Type

/**
 * V3 Emotion Health Reminder Param Model
 */
open class IDOEmotionHealthReminderParamModel(
    operate: Int,
    notifyType: Int = 0,
    emotionalHealthSwitch: Int,
    pressureRemind: IDOPressureRemindConfig? = null,
    unpleasantRemind: IDOUnpleasantRemindConfig? = null
) : IDOBaseAdapterModel<IDOEmotionHealthReminderParamModel> {
    /** Protocol version number */
    @SerializedName("version")
    private val version: Int = 4

    /** Operation type: 0:Invalid 1:Set 2:Query */
    @SerializedName("operate")
    var operate: Int = operate

    /** Notification type: 0:Invalid 1:Allow notification 2:Silent notification 3:Close notification */
    @SerializedName("notify_type")
    var notifyType: Int = notifyType

    /** Emotional health master switch: 1:On 0:Off */
    @SerializedName("emotional_health_switch")
    var emotionalHealthSwitch: Int = emotionalHealthSwitch

    /** Pressure reminder configuration */
    @SerializedName("pressure_remind")
    var pressureRemind: IDOPressureRemindConfig? = pressureRemind

    /** Unpleasant reminder configuration */
    @SerializedName("unpleasant_remind")
    var unpleasantRemind: IDOUnpleasantRemindConfig? = unpleasantRemind

    override fun getDeSerializer(): IDOBaseModelDeSerializer<IDOEmotionHealthReminderParamModel>? {
        return EmotionHealthReminderParamModelSerializer()
    }
}

internal class EmotionHealthReminderParamModelSerializer : IDOBaseModelDeSerializer<IDOEmotionHealthReminderParamModel> {
    override fun serialize(src: IDOEmotionHealthReminderParamModel, typeOfSrc: Type, context: JsonSerializationContext): JsonElement {
        val jsonObject = JsonObject()
        jsonObject.addProperty("version", 4)
        jsonObject.addProperty("operate", src.operate)
        jsonObject.addProperty("notify_type", src.notifyType)
        jsonObject.addProperty("emotional_health_switch", src.emotionalHealthSwitch)

        if (src.pressureRemind != null) {
            val pressureSerializer = PressureRemindConfigSerializer()
            jsonObject.add("pressure_remind", pressureSerializer.serialize(src.pressureRemind!!, IDOPressureRemindConfig::class.java, context))
        }

        if (src.unpleasantRemind != null) {
            jsonObject.add("unpleasant_remind", context.serialize(src.unpleasantRemind))
        }

        return jsonObject
    }

    override fun deserialize(json: JsonElement?, typeOfT: Type?, context: JsonDeserializationContext?): IDOEmotionHealthReminderParamModel {
        json?.let {
            val obj = it.asJsonObject
            val operate = obj.get("operate")?.asInt ?: 0
            val notifyType = obj.get("notify_type")?.asInt ?: 0
            val emotionalHealthSwitch = obj.get("emotional_health_switch")?.asInt ?: 0

            var pressureRemind: IDOPressureRemindConfig? = null
            if (obj.has("pressure_remind")) {
                val pressureSerializer = PressureRemindConfigSerializer()
                pressureRemind = pressureSerializer.deserialize(obj.get("pressure_remind"), IDOPressureRemindConfig::class.java, context)
            }

            var unpleasantRemind: IDOUnpleasantRemindConfig? = null
            if (obj.has("unpleasant_remind")) {
                unpleasantRemind = context!!.deserialize(obj.get("unpleasant_remind"), IDOUnpleasantRemindConfig::class.java)
            }

            return IDOEmotionHealthReminderParamModel(operate, notifyType, emotionalHealthSwitch, pressureRemind, unpleasantRemind)
        }
        return IDOEmotionHealthReminderParamModel(0, 0, 0)
    }
}

/**
 * Pressure reminder configuration
 */
open class IDOPressureRemindConfig(
    pressureSwitch: Int = 0,
    startHour: Int = 0,
    startMinute: Int = 0,
    endHour: Int = 0,
    endMinute: Int = 0,
    repeats: HashSet<IDOWeek> = HashSet(),
    reminderInterval: Int = 0
) : Serializable {
    /** Pressure reminder switch: 1:On 0:Off */
    var pressureSwitch: Int = pressureSwitch

    /** Reminder start time (hour) */
    var startHour: Int = startHour

    /** Reminder start time (minute) */
    var startMinute: Int = startMinute

    /** Reminder end time (hour) */
    var endHour: Int = endHour

    /** Reminder end time (minute) */
    var endMinute: Int = endMinute


    /**
     * Repeat cycle (week days)
     * Using IDOWeek enum: MONDAY(0), TUESDAY(1), ..., SUNDAY(6)
     */
    var repeats: HashSet<IDOWeek> = repeats

    /** Reminder interval, unit: minutes */
    var reminderInterval: Int = reminderInterval
}

/**
 * Serializer & Deserializer for IDOPressureRemindConfig
 */
internal class PressureRemindConfigSerializer : JsonSerializer<IDOPressureRemindConfig>, JsonDeserializer<IDOPressureRemindConfig> {
    override fun serialize(src: IDOPressureRemindConfig, typeOfSrc: Type, context: JsonSerializationContext): JsonElement {
        val jsonObject = JsonObject()
        jsonObject.addProperty("pressure_switch", src.pressureSwitch)
        jsonObject.addProperty("start_hour", src.startHour)
        jsonObject.addProperty("start_minute", src.startMinute)
        jsonObject.addProperty("end_hour", src.endHour)
        jsonObject.addProperty("end_minute", src.endMinute)
        jsonObject.addProperty("reminder_interval", src.reminderInterval)

        // Encode repetitions: bit0 is master switch, bit1~bit7 are Monday to Sunday
        var repetitions = 0
        // repetitions = repetitions or ((if (src.isOpenRepeat) 1 else 0) shl 0) // bit0 reserved/not used in this version

        src.repeats.forEach { week ->
            repetitions = repetitions or (1 shl (week.rawValue + 1))
        }
        jsonObject.addProperty("repetitions", repetitions)

        return jsonObject
    }

    override fun deserialize(json: JsonElement?, typeOfT: Type?, context: JsonDeserializationContext?): IDOPressureRemindConfig {
        json?.let {
            val jsonObject = it.asJsonObject
            val pressureSwitch = jsonObject.get("pressure_switch")?.asInt ?: 0
            val startHour = jsonObject.get("start_hour")?.asInt ?: 0
            val startMinute = jsonObject.get("start_minute")?.asInt ?: 0
            val endHour = jsonObject.get("end_hour")?.asInt ?: 0
            val endMinute = jsonObject.get("end_minute")?.asInt ?: 0
            val reminderInterval = jsonObject.get("reminder_interval")?.asInt ?: 0

            // Decode repetitions: bit0 is master switch, bit1~bit7 are Monday to Sunday
            val repetitions = jsonObject.get("repetitions")?.asInt ?: 0
            // val isOpenRepeat = (repetitions and (1 shl 0)) != 0
            val repeats = HashSet<IDOWeek>()
            for (i in 0..6) {
                if ((repetitions and (1 shl (i + 1))) != 0) {
                    repeats.add(IDOWeek.fromRawValue(i))
                }
            }

            return IDOPressureRemindConfig(
                pressureSwitch,
                startHour,
                startMinute,
                endHour,
                endMinute,
                repeats,
                reminderInterval
            )
        }
        return IDOPressureRemindConfig()
    }
}

/**
 * Unpleasant reminder configuration
 */
open class IDOUnpleasantRemindConfig(
    unpleasantSwitch: Int = 0,
    unhappyReminderNum: Int = 0
) : Serializable {
    /** Unpleasant reminder switch: 1:On 0:Off */
    @SerializedName("unpleasant_switch")
    var unpleasantSwitch: Int = unpleasantSwitch

    /** Unhappy reminder threshold, 0 is invalid */
    @SerializedName("unhappy_reminder_num")
    var unhappyReminderNum: Int = unhappyReminderNum
}

/**
 * V3 Emotion Health Reminder Reply Model
 */
open class IDOEmotionHealthReminderReplyModel(
    operate: Int = 0,
    errorCode: Int = 0,
    notifyType: Int = 0,
    emotionalHealthSwitch: Int = 0,
    pressureRemind: IDOPressureRemindConfig? = null,
    unpleasantRemind: IDOUnpleasantRemindConfig? = null
) : IDOBaseAdapterModel<IDOEmotionHealthReminderReplyModel> {
    /** Protocol version number */
    @SerializedName("version")
    private val version: Int = 0

    /** Operation type: 0:Invalid 1:Set 2:Query */
    @SerializedName("operate")
    var operate: Int = operate

    /** Error code: 0:Success, non-zero is error code */
    @SerializedName("error_code")
    var errorCode: Int = errorCode

    /** Notification type: 0:Invalid 1:Allow notification 2:Silent notification 3:Close notification */
    @SerializedName("notify_type")
    var notifyType: Int = notifyType

    /** Emotional health master switch: 1:On 0:Off */
    @SerializedName("emotional_health_switch")
    var emotionalHealthSwitch: Int = emotionalHealthSwitch

    /** Pressure reminder configuration */
    @SerializedName("pressure_remind")
    var pressureRemind: IDOPressureRemindConfig? = pressureRemind

    /** Unpleasant reminder configuration */
    @SerializedName("unpleasant_remind")
    var unpleasantRemind: IDOUnpleasantRemindConfig? = unpleasantRemind

    override fun getDeSerializer(): IDOBaseModelDeSerializer<IDOEmotionHealthReminderReplyModel>? {
        return EmotionHealthReminderReplyModelSerializer()
    }
}

internal class EmotionHealthReminderReplyModelSerializer : IDOBaseModelDeSerializer<IDOEmotionHealthReminderReplyModel> {
    override fun serialize(src: IDOEmotionHealthReminderReplyModel, typeOfSrc: Type, context: JsonSerializationContext): JsonElement {
        val jsonObject = JsonObject()
        jsonObject.addProperty("version", 0)
        jsonObject.addProperty("operate", src.operate)
        jsonObject.addProperty("error_code", src.errorCode)
        jsonObject.addProperty("notify_type", src.notifyType)
        jsonObject.addProperty("emotional_health_switch", src.emotionalHealthSwitch)

        if (src.pressureRemind != null) {
            val pressureSerializer = PressureRemindConfigSerializer()
            jsonObject.add("pressure_remind", pressureSerializer.serialize(src.pressureRemind!!, IDOPressureRemindConfig::class.java, context))
        }

        if (src.unpleasantRemind != null) {
            jsonObject.add("unpleasant_remind", context.serialize(src.unpleasantRemind))
        }

        return jsonObject
    }

    override fun deserialize(json: JsonElement?, typeOfT: Type?, context: JsonDeserializationContext?): IDOEmotionHealthReminderReplyModel {
        json?.let {
            val obj = it.asJsonObject
            val operate = obj.get("operate")?.asInt ?: 0
            val errorCode = obj.get("error_code")?.asInt ?: 0
            val notifyType = obj.get("notify_type")?.asInt ?: 0
            val emotionalHealthSwitch = obj.get("emotional_health_switch")?.asInt ?: 0

            var pressureRemind: IDOPressureRemindConfig? = null
            if (obj.has("pressure_remind")) {
                val pressureSerializer = PressureRemindConfigSerializer()
                pressureRemind = pressureSerializer.deserialize(obj.get("pressure_remind"), IDOPressureRemindConfig::class.java, context)
            }

            var unpleasantRemind: IDOUnpleasantRemindConfig? = null
            if (obj.has("unpleasant_remind")) {
                unpleasantRemind = context!!.deserialize(obj.get("unpleasant_remind"), IDOUnpleasantRemindConfig::class.java)
            }

            return IDOEmotionHealthReminderReplyModel(operate, errorCode, notifyType, emotionalHealthSwitch, pressureRemind, unpleasantRemind)
        }
        return IDOEmotionHealthReminderReplyModel()
    }
}

File: ido_sdk_flutter/ido_sdk/protocol_sdk/protocol_channel/ios/Classes/Model/IDOEmotionHealthReminderModel.swift
//
//  IDOEmotionHealthReminderModel.swift
//  protocol_channel
//
//  Created by hc on 2025/12/04.
//

import Foundation

// V3 Emotion Health Reminder Param Model
@objcMembers
public class IDOEmotionHealthReminderParamModel: NSObject, IDOBaseModel {
    /// Protocol version number
    private var version: Int
    
    /// Operation type: 0:Invalid 1:Set 2:Query
    public var operate: Int
    
    /// Notification type: 0:Invalid 1:Allow notification 2:Silent notification 3:Close notification
    public var notifyType: Int
    
    /// Emotional health master switch: 1:On 0:Off
    public var emotionalHealthSwitch: Int
    
    /// Pressure reminder configuration
    public var pressureRemind: IDOPressureRemindConfig?
    
    /// Unpleasant reminder configuration
    public var unpleasantRemind: IDOUnpleasantRemindConfig?
    
    enum CodingKeys: String, CodingKey {
        case version = "version"
        case operate = "operate"
        case notifyType = "notify_type"
        case emotionalHealthSwitch = "emotional_health_switch"
        case pressureRemind = "pressure_remind"
        case unpleasantRemind = "unpleasant_remind"
    }
    
    public init(operate: Int, notifyType: Int = 0, emotionalHealthSwitch: Int, pressureRemind: IDOPressureRemindConfig? = nil, unpleasantRemind: IDOUnpleasantRemindConfig? = nil) {
        self.version = 4
        self.operate = operate
        self.notifyType = notifyType
        self.emotionalHealthSwitch = emotionalHealthSwitch
        self.pressureRemind = pressureRemind
        self.unpleasantRemind = unpleasantRemind
    }
    
    public func toJsonString() -> String? {
        if let jsonData = try? JSONEncoder().encode(self) {
            let jsonString = String(data: jsonData, encoding: .utf8)
            return jsonString
        }
        return nil
    }
}

// MARK: - Pressure Remind Config

/// Pressure reminder configuration
public class IDOPressureRemindConfig: NSObject, Codable {
    /// Pressure reminder switch: 1:On 0:Off
    public var pressureSwitch: Int
    
    /// Reminder start time (hour)
    public var startHour: Int
    
    /// Reminder start time (minute)
    public var startMinute: Int
    
    /// Reminder end time (hour)
    public var endHour: Int
    
    /// Reminder end time (minute)
    public var endMinute: Int
    
    /// Repeat cycle
    public var repeats = Set<IDOWeek>()
    
    // /// Repeat On/Off
    // public var isOpenRepeat: Bool
    
    /// Reminder interval, unit: minutes
    public var reminderInterval: Int
    
    enum CodingKeys: String, CodingKey {
        case pressureSwitch = "pressure_switch"
        case startHour = "start_hour"
        case startMinute = "start_minute"
        case endHour = "end_hour"
        case endMinute = "end_minute"
        case repetitions = "repetitions"
        case reminderInterval = "reminder_interval"
    }
    
    public init(pressureSwitch: Int, startHour: Int, startMinute: Int, endHour: Int, endMinute: Int, repeats: Set<IDOWeek>, reminderInterval: Int) {
        self.pressureSwitch = pressureSwitch
        self.startHour = startHour
        self.startMinute = startMinute
        self.endHour = endHour
        self.endMinute = endMinute
        self.repeats = repeats
        // self.isOpenRepeat = isOpenRepeat
        self.reminderInterval = reminderInterval
    }
    
    required public init(from decoder: Decoder) throws {
        let container = try decoder.container(keyedBy: CodingKeys.self)
        pressureSwitch = try container.decode(Int.self, forKey: .pressureSwitch)
        startHour = try container.decode(Int.self, forKey: .startHour)
        startMinute = try container.decode(Int.self, forKey: .startMinute)
        endHour = try container.decode(Int.self, forKey: .endHour)
        endMinute = try container.decode(Int.self, forKey: .endMinute)
        reminderInterval = try container.decode(Int.self, forKey: .reminderInterval)
        
        // Decode repetitions: bit0 is master switch, bit1~bit7 are Monday to Sunday
        if let value = try container.decodeIfPresent(Int.self, forKey: .repetitions) {
            // isOpenRepeat = (value & (1 << 0) != 0)
            var items = Set<IDOWeek>()
            for i in 0...6 {
                if value & (1 << (i + 1)) != 0 {
                    items.insert(IDOWeek(rawValue: i)!)
                }
            }
            repeats = items
        } else {
            //isOpenRepeat = false
            repeats = Set<IDOWeek>()
        }
    }
    
    public func encode(to encoder: Encoder) throws {
        var container = encoder.container(keyedBy: CodingKeys.self)
        try container.encode(pressureSwitch, forKey: .pressureSwitch)
        try container.encode(startHour, forKey: .startHour)
        try container.encode(startMinute, forKey: .startMinute)
        try container.encode(endHour, forKey: .endHour)
        try container.encode(endMinute, forKey: .endMinute)
        try container.encode(reminderInterval, forKey: .reminderInterval)
        
        // Encode repetitions: bit0 is master switch, bit1~bit7 are Monday to Sunday
        var value = 0
        //value |= (isOpenRepeat ? 1 : 0) << 0 // bit0为预留
        for item in repeats {
            value |= (1 << (item.rawValue + 1))
        }
        try container.encode(value, forKey: .repetitions)
    }
}

// MARK: - Unpleasant Remind Config

/// Unpleasant reminder configuration
public class IDOUnpleasantRemindConfig: NSObject, Codable {
    /// Unpleasant reminder switch: 1:On 0:Off
    public var unpleasantSwitch: Int
    
    /// Unhappy reminder threshold, 0 is invalid
    public var unhappyReminderNum: Int
    
    enum CodingKeys: String, CodingKey {
        case unpleasantSwitch = "unpleasant_switch"
        case unhappyReminderNum = "unhappy_reminder_num"
    }
    
    public init(unpleasantSwitch: Int, unhappyReminderNum: Int) {
        self.unpleasantSwitch = unpleasantSwitch
        self.unhappyReminderNum = unhappyReminderNum
    }
}

// MARK: - Reply Model

/// V3 Emotion Health Reminder Reply Model
@objcMembers
public class IDOEmotionHealthReminderReplyModel: NSObject, IDOBaseModel {
    /// Protocol version number
    private var version: Int
    
    /// Operation type: 0:Invalid 1:Set 2:Query
    public var operate: Int
    
    /// Error code: 0:Success, non-zero is error code
    public var errorCode: Int
    
    /// Notification type: 0:Invalid 1:Allow notification 2:Silent notification 3:Close notification
    public var notifyType: Int
    
    /// Emotional health master switch: 1:On 0:Off
    public var emotionalHealthSwitch: Int
    
    /// Pressure reminder configuration
    public var pressureRemind: IDOPressureRemindConfig?
    
    /// Unpleasant reminder configuration
    public var unpleasantRemind: IDOUnpleasantRemindConfig?
    
    enum CodingKeys: String, CodingKey {
        case version = "version"
        case operate = "operate"
        case errorCode = "error_code"
        case notifyType = "notify_type"
        case emotionalHealthSwitch = "emotional_health_switch"
        case pressureRemind = "pressure_remind"
        case unpleasantRemind = "unpleasant_remind"
    }
    
    public init(operate: Int, errorCode: Int, notifyType: Int = 0, emotionalHealthSwitch: Int = 0, pressureRemind: IDOPressureRemindConfig? = nil, unpleasantRemind: IDOUnpleasantRemindConfig? = nil) {
        self.version = 0
        self.operate = operate
        self.errorCode = errorCode
        self.notifyType = notifyType
        self.emotionalHealthSwitch = emotionalHealthSwitch
        self.pressureRemind = pressureRemind
        self.unpleasantRemind = unpleasantRemind
    }
    
    public func toJsonString() -> String? {
        if let jsonData = try? JSONEncoder().encode(self) {
            let jsonString = String(data: jsonData, encoding: .utf8)
            return jsonString
        }
        return nil
    }
}
